### ISSUE TITLE
How to fix "Kernel#open is a serious security risk" warnings

### ISSUE THREAD

**Issue by huyvohcmc** [rubocop-hq/rubocop#6216]
With the logic like this:
```ruby
require 'open-uri'

document_url = a file on the internet
content = open(document_url).read
```
Rubocop will raise the `Security/Open: The use of Kernel#open is a serious security risk` warning. I look up to the [documentation](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Security/Open) and it suggests `File.open(something)` or `IO.popen(something)` but neither of them works when opening files on the internet.

How can I get on with it?

**Comment by kamok** [comment]
Would be helpful to tell people to use Net:HTTP too should they use open-uri for that.

**Comment by joshuapinter** [comment]
Can we get some clarity here. 

Should we not use `URI.parse(something).open` anymore?

Is `Net::HTTP.get_response(URI(url))` favoured now?

Thanks.

**Comment by diaclavijo** [comment]
> With the logic like this:
> 
> ```ruby
> require 'open-uri'
> 
> document_url = a file on the internet
> content = open(document_url).read
> ```
> 
> Rubocop will raise the `Security/Open: The use of Kernel#open is a serious security risk` warning. I look up to the [documentation](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Security/Open) and it suggests `File.open(something)` or `IO.popen(something)` but neither of them works when opening files on the internet.
> 
> How can I get on with it?

This cop checks for the use of `Kernel#open` and `URI.open`.

`Kernel#open` and `URI.open` enable not only file access but also process invocation by prefixing a pipe symbol (e.g., `open(“| ls”)`). So, it may lead to a serious security risk by using variable input to the argument of `Kernel#open` and `URI.open`. It would be better to use `File.open`, `IO.popen` or `URI.parse#open` explicitly.

https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Security/Open

**Comment by stefanhorning** [comment]
Actually just use `URI.open()` which works as a drop in replacement for `(Kernel.)open` and will be accepted by Rubocop too, in contrary what the docs say (RC version 0.93.1).

Actually `URI.parse(something).open` doesn't exist (tested with ruby 2.7).

**Comment by OlivierJM** [comment]
> Actually just use `URI.open()` which works as a drop in replacement for `Kernel.open()` (i.e. `open()`) and will be accepted by Rubocop too, in contrary what the docs say (RC version 0.93.1).
> 
> Actually `URI.parse(something).open` doesn't exist (tested with ruby 2.7).

I replaced `open()` with `URI.open()` and rubocop smiled.

**Comment by terrencec05** [comment]
> > Actually just use `URI.open()` which works as a drop in replacement for `Kernel.open()` (i.e. `open()`) and will be accepted by Rubocop too, in contrary what the docs say (RC version 0.93.1).
> > Actually `URI.parse(something).open` doesn't exist (tested with ruby 2.7).
> 
> I replaced `open()` with `URI.open()` and rubocop smiled.

rubocop is happier if we use `File.open()`

**Comment by sandipsubedi** [comment]
> URI.parse(something).open

^ worked for me

**Comment by SaitoJP** [comment]
> Actually just use `URI.open()` which works as a drop in replacement for `Kernel.open()` (i.e. `open()`) and will be accepted by Rubocop too, in contrary what the docs say (RC version 0.93.1).
> 
> Actually `URI.parse(something).open` doesn't exist (tested with ruby 2.7).

URI.parse(something).open` doesn't exist in ruby 2.7, but it does in Rails.

**Comment by channainfo** [comment]
I have the same problem in ruby 2.7 
> URI.parse(something).open

returns the following error:

```

NoMethodError:
       private method `open' called for #<URI::Generic:0x00007fcc03e89640>
```

**Comment by harry-wood** [comment]
The folks who are saying `URI.open()` works are presumably using an explicit URL string e.g rubocop is fine with `URI.open('https://myurl.com')`  but with a variable e.g. `URI.open(some_variable)` rubocop says _"security risk"_.

`URI.parse(some_variable).open` works, BUT you need to do a `require 'open-uri'` or use rails or nokogiri or something else which brings open-uri in. Testing it with only ruby (v2.7 or v3.1) in a fresh irb session, we get the _"private method 'open' called"_ error. It seems open-uri replaces that private "open" method with [this public "open" method](https://ruby-doc.org/stdlib-2.6.3/libdoc/open-uri/rdoc/URI.html), and rubocop is happy with passing a variable into that.

Is there a good recommendation which does not need `require 'open-uri'` and where rubocop doesn't mind the variable? I guess the answer is "not really" because, you'd get into using `Net::HTTP` with all the wrapping code you need to use that properly. Anyway never versions of ruby also need `require 'uri'`!

**Comment by mzrnsh** [comment]
@harry-wood Good catch. As a quick fix I kept the protocol fragment from my URL to keep Rubocop happy.

So instead of ~~`URI.open(some_variable)`~~ I went with `URI.open("https://#{some_variable}")`. Obviously, it won't work if you are not 100% sure about which protocol to expect.

**Comment by dani24gl** [comment]
What I actually needed is the content  of the image file, so I used `URI.parse(url).read` and it worked (the difference is that I'm using **read** instead of **open**)

**Comment by darkthirst** [comment]
hi

I couldn't find any documentation on around how to fix this:

`URI.open(full_url, "Api-Key" => messenger_api_key)`

**Comment by dorianmariecom** [comment]
I did `URI.parse(url).open.read`
